src/ucp/core/ucp_listener.c

```diff
@@ -42,44 +42,73 @@ void ucp_listener_schedule_accept_cb(ucp_ep_h ep)
 
 static unsigned ucp_listener_conn_request_progress(void *arg)
 {
-    ucp_listener_accept_t *accept = arg;
-    ucp_ep_h ep = accept->ep;
-    ucs_status_t status;
+    ucp_listener_accept_t            *accept      = arg;
+    const ucp_wireup_sockaddr_priv_t *client_data = accept->wireup_data;
+    ucp_listener_h                   listener     = accept->listener;
+    ucp_unpacked_address_t           client_address;
+    ucp_ep_params_t                  params;
+    ucp_ep_h                         ep;
+    ucs_status_t                     status;
 
-    ucs_trace_func("listener=%p ep=%p", accept->listener, ep);
+    ucs_trace_func("listener=%p", listener);
 
-    if (!(ep->flags & UCP_EP_FLAG_LISTENER)) {
-        /* send wireup request message, to connect the client to the server's new endpoint */
-        ucs_assert(!(ep->flags & UCP_EP_FLAG_CONNECT_REQ_QUEUED));
-        status = ucp_wireup_send_request(ep);
-        if (status != UCS_OK) {
-            goto err_destroy_ep;
+    status = ucp_address_unpack(client_data + 1, &client_address);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    params.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    params.err_mode   = client_data->err_mode;
+
+    if (client_data->is_full_addr) {
+        /* create endpoint to the worker address we got in the private data */
+        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
+                                              &client_address,
+                                              UCP_EP_CREATE_AM_LANE, "listener",
+                                              &ep);
+        if (status == UCS_OK) {
+            ucp_ep_update_dest_ep_ptr(ep, client_data->ep_ptr);
+
+            /* send wireup request message, to connect the client to the server's new endpoint */
+            ucs_assert(!(ep->flags & UCP_EP_FLAG_CONNECT_REQ_QUEUED));
+            if ((status = ucp_wireup_send_request(ep)) != UCS_OK) {
```

Comment:
IMHO it would be better to invoke the function outside the if, like it's done everywhere else...
same for line 86.